datalake/src/test/java/com/sequenceiq/datalake/flow/upgrade/ccm/UpgradeCcmActionsTest.java

```diff
@@ -0,0 +1,290 @@
+
+        testErrorInSdxStatusService(underTest::finishedAction);
+        verify(sdxStatusService).setStatusForDatalakeAndNotify(DatalakeStatusEnum.RUNNING, "Datalake is running", SDX_ID);
+        UpgradeCcmSuccessEvent payload = (UpgradeCcmSuccessEvent) payloadArgumentCaptor.getValue();
+        assertThat(payload.getResourceId()).isEqualTo(SDX_ID);
+        assertThat(payload.getUserId()).isEqualTo(USER_ID);
+    }
+
+    @Test
+    void failureAction() {
+        IllegalStateException failureException = new IllegalStateException(MESSAGE);
+        actionPayload = new UpgradeCcmFailedEvent(SDX_ID, USER_ID, failureException);
+        setupContextWithPayload();
+        testUpgradeActionHappyPath(underTest::failedAction);
+        UpgradeCcmFailedEvent payload = (UpgradeCcmFailedEvent) payloadArgumentCaptor.getValue();
+        assertThat(selectorArgumentCaptor.getValue()).isEqualTo(UPGRADE_CCM_FAILED_HANDLED_EVENT.event());
+        assertThat(payload.getResourceId()).isEqualTo(SDX_ID);
+        assertThat(payload.getUserId()).isEqualTo(USER_ID);
+        verify(sdxStatusService).setStatusForDatalakeAndNotify(DatalakeStatusEnum.DATALAKE_UPGRADE_CCM_FAILED, MESSAGE, SDX_ID);
+        verify(metricService).incrementMetricCounter(eq(MetricType.UPGRADE_CCM_FAILED), (SdxCluster) any());
+    }
+
+    private Action<?, ?> configureAction(Supplier<Action<?, ?>> actionSupplier) {
+        Action<?, ?> action = actionSupplier.get();
+        assertThat(action).isNotNull();
+        setActionPrivateFields(action);
+        AbstractAction abstractAction = (AbstractAction) action;
+        abstractAction.setFailureEvent(failureEvent);
+        return action;
+    }
+
+    private void setActionPrivateFields(Action<?, ?> action) {
+        ReflectionTestUtils.setField(action, null, runningFlows, FlowRegister.class);
+        ReflectionTestUtils.setField(action, null, eventBus, EventBus.class);
+        ReflectionTestUtils.setField(action, null, reactorEventFactory, ErrorHandlerAwareReactorEventFactory.class);
+        ReflectionTestUtils.setField(action, null, tracer, Tracer.class);
+    }
+
+    private void verifyFailureEvent() {
+        assertThat(selectorArgumentCaptor.getValue()).isEqualTo(FAILURE_EVENT);
+
+        verifyEventFactoryAndHeaders();
+
+        assertThat(payloadArgumentCaptor.getValue()).isSameAs(actionPayload);
+    }
+
+    private void verifyEventFactoryAndHeaders() {
+        assertThat(eventArgumentCaptor.getValue()).isSameAs(event);
+
+        Map<String, Object> headers = headersArgumentCaptor.getValue();
+        assertThat(headers).isNotNull();
+        assertThat(headers.get(FlowConstants.FLOW_ID)).isEqualTo(FLOW_ID);
+        assertThat(headers.get(FlowConstants.FLOW_TRIGGER_USERCRN)).isEqualTo(FLOW_TRIGGER_USER_CRN);
+    }
+
+    private void testUpgradeActionHappyPath(Supplier<Action<?, ?>> upgradeAction) {
+        Action<?, ?> action = configureAction(upgradeAction);
+        action.execute(stateContext);
```

Comment:
This certainly works, but here you are implicitly testing the behavior of `AbstractAction` itself, thus a lot of extra mocking is needed. In case you want to verify only the action `protected` methods, there is `AbstractActionTestSupport` in module `flow` (test scope). Up to you if you go down the simpler (& lazier) or the harder (& more complete) route. :smile:

Reply:
using it would save only a few lines so I would skip it for now